include/onnxruntime/core/graph/graph.h

```diff
@@ -636,8 +636,12 @@ class Graph {
 #endif  // !defined(ORT_MINIMAL_BUILD)
 
 #if !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)
-  /** Add an initializer tensor to the Graph. */
-  void AddInitializedTensor(const ONNX_NAMESPACE::TensorProto& tensor_proto);
+  /** Add an initializer tensor to the Graph. 
+    @param force_nodearg_creation If true creates a corresponding NodeArg for the initialized tensor.
+    If false, a NodeArg is created ONLY if the graph is not loaded from a model file and a NodeArg
+    with the same name as the initialized tensor_proto does not exist.
+  */
+  void AddInitializedTensor(const ONNX_NAMESPACE::TensorProto& tensor_proto, bool force_nodearg_creation = false);
```

Comment:
When would clients of this class want to force the creation of node arg? In other words, under what circumstances should I set this to true?

Reply:
Well clients need to ensure that a NodeArg exists for an initialized tensor that is getting added to the graph. Either, they do a `GetOrCreateNodeArg()` on the client side along with `AddInitializedTensor()` or they do a `AddInitializedTensor(..., true)` and force the initializer addition logic to create a NodeArg (if it doesn't already exist). This new interface is just exposing a new way to keep client side code a bit more concise. 

Some optimizers are lucky enough to avoid creating a NodeArg altogether for an initialized tensor because they will use a name for the initializer that already exists in the graph - for example, Constant Folding, where the name of the initializer is the same as the output of a node getting folded in which case a separate NodeArg creation step is not required.

Reply:
I am curious why this was specifically NOT created in the first place? This change may be circumventing some prior logic. The PR description does not address it.

Reply:
NOT creating it in the first place is the oversight/bug that this PR is solving :)

Perhaps you can elaborate on what you think the PR may be circumventing ? I thought the PR description was clear enough but I am happy to improve the description if you have something more specific.

FWIW - I spoke to the original implementer of this fusion @skottmckay and we did agree that this was a bug and he was the one who recommended this interface change in `AddInitializedTensor()`. In my original fix, I had the client-side optimizer create the NodeArg (as opposed to having `AddInitializedTensor()` create the NodeArg).

Reply:
Should clients be relieved from the responsibility of setting this boolean? Can AddInitializedTensor take this decision on behalf of the clients just like how it's taking a decision based on is_loaded_from_model_file_?

Reply:
Not sure if it can/supposed to take such a decision. 

The clients can never truly be "relieved" of this responsibility (they never were / they never are going to be). The client needs to make sure that the corresponding NodeArg exists. Even in existing optimizers, there is a call to GetOrCreateNodeArg() where applicable. This is exactly the same but via a boolean flag.

Reply:
The semantics of the function seem a bit inconsistent. In one case it takes the decision on behalf of the client and in other cases it expects the client to tell it explicitly to create the NodeArg. AddInitializedTensor should either assume that the NodeArg is created or create it if not present. I don't think there's any use case of calling AddInitializedTensor where you wouldn't create a NodeArg for it. We can take this offline.

Reply:
Spoke offline. Will remove the change to AddInitializedTensor(). Will keep all the changes local to the optimizer.